### ISSUE TITLE
A few rules are mostly redundant in newer TypeScripts with specific compiler options

### ISSUE THREAD

**Issue by isiahmeadows** [palantir/tslint#1861]
Here's the relevant rules of varying redundancy unaddressed in the docs:

- `no-invalid-this` &rarr; `--noImplicitThis` (not a precise equivalent, but it makes all usages explicit)
- `no-switch-case-fallthrough` &rarr; `--noFallthroughCasesInSwitch` (exact equivalent)
- `no-unused-variable` &rarr; `--noUnusedLocals` and `--noUnusedParameters` (exact equivalent between the two)
- `switch-default` is partially redundant with `--noImplicitReturns`, which requires that all code paths are reachable *and* return a value. (It only checks the end, but a final switch-case is rather common.)

**Comment by pe8ter** [comment]
If you deprecate options like `no-unused-variable` in favor of TypeScript compiler flags then you can't opt out with rule flags like `// tslint:disable-next-line`.

**Comment by jsynowiec** [comment]
There is also an issue with `--noUnusedLocals` and `--noUnusedParameters` triggering when function signature has several parameters but you only use the last one, the first few are unneeded (but still must be declared).

For e.g.
```js
function foo(a: any, b: T): T { return b; }
```

See https://github.com/Microsoft/TypeScript/issues/9458 for reference.

The *approved fix* is to prefix uninteresting parameters with an underscore, see https://github.com/Microsoft/TypeScript/pull/9464 which breaks the `variable-name` TSLint rule.

**Comment by ajafff** [comment]
@jsynowiec If you need fine grained control for variable name formats, I would recommend one of my custom rules https://github.com/ajafff/tslint-consistent-codestyle#naming-convention
If you scroll down a bit you get to the Examples section where you should find what you need. If you have trouble setting it up for your needs, just leave me an issue over there.

Btw. sorry for shamelessly advertising my own package here.

**Comment by pdc** [comment]
I don’t want to relax the naming rules except in this case. It seems we need a new rule which checks that function parameters start with an underscore if and only if they are (a) unused and (b) followed by a parameter that *is* used. Slightly annoying. :-)

**Comment by isiahmeadows** [comment]
So, now that I see, I'm thinking:

- `no-unused-variable` should be kept. (It makes prototyping on existing code bases easier, since you can just disable it for the unused parameter in question. I personally use the ESLint equivalent frequently)
- `no-invalid-this` is fully unnecessary. The correct opt out is just by declaring the type of `this`, like in `function foo(this: any, ...) { ... }`.
- `no-switch-case-fallthrough` *really* needs a compiler opt-out, since fallthrough can really avoid code duplication when done right.
- `switch-default` should still be kept, since the compiler option isn't a complete replacement. But it'd be worth noting in the docs.

Although, here's a highly relevant TS issue: https://github.com/Microsoft/TypeScript/issues/11051

**Comment by filipesilva** [comment]
Heya, I just wanted to pitch in in favor of keeping `no-unused-variable`. 

My argument is similar to the value of prototyping one. I think it's a different 'level' of checking. You might want a compilation error... or you might want a lint check before commiting. 

You might also want to treat different parts of your app with different linting rules, without having a different tsconfig.json.

So I think there's value in having both a linting rule and a compilation setting.

**Comment by IllusionMH** [comment]
As I understand - there rules won't be used in `tslint:recommended` configuration and will be removed from this repo.

Maybe it worth to move them to separate repository/package, provide "migration guide" on how to include and enable them in users configuration for those (I expect it to happen on major release)?

**Comment by andy-hanson** [comment]
Could we deprecate (except `no-unused-variable`) these for 5.0?

**Comment by adidahiya** [comment]
@andy-hanson yeah let's do it

**Comment by nchen63** [comment]
keeping these rules for now. See https://github.com/palantir/tslint/pull/2422#issuecomment-289879986

**Comment by glen-84** [comment]
Should this be revisited now that https://github.com/Microsoft/TypeScript/issues/9448 was fixed, or wait for https://github.com/Microsoft/TypeScript/issues/19139?

I suppose since they require a specific version of TypeScript, the rules should remain until TSLint stops supporting older versions of TypeScript (where they would still be useful).

**Comment by isiahmeadows** [comment]
If this helps make that judgement, the version `@ts-ignore` first appeared (2.6) was released late October 2017, and the current version (2.9) was released late May. Currently, what TSLint officially supports is [2.1 or later](https://github.com/palantir/tslint/blob/master/package.json#L54), but 2.1 was released December of 2016.
